File: src/dataset/cpp/simple-linked-list_exercism-toemm/gemini/GeminiFlash_simple_linked_list_v2.cpp
#include "simple_linked_list.h"

#include <stdexcept>
#include <algorithm> // For std::swap

namespace simple_linked_list {

std::size_t List::size() const {
    return current_size;
}

void List::push(int entry) {
    Element* new_head = new Element{entry, head};
    head = new_head;
    current_size++;
}

int List::pop() {
    if (head == nullptr) {
        throw std::domain_error("no elements");
    }

    Element* old_head = head;
    int data = old_head->data;
    head = old_head->next;
    delete old_head;
    current_size--;
    return data;
}

void List::reverse() {
    if (current_size <= 1) return;

    Element* prev = nullptr;
    Element* current = head;
    Element* next = nullptr;

    while (current != nullptr) {
        next = current->next;
        current->next = prev;
        prev = current;
        current = next;
    }

    head = prev;
}


List::~List() {
    Element* current = head;
    while (current != nullptr) {
        Element* next = current->next;
        delete current;
        current = next;
    }
}

}  // namespace simple_linked_list

File: src/dataset/Java/simple-linked-list_exercism-uzilan/gemini/GeminiFlash_SimpleLinkedList_v2.java
import java.lang.reflect.Array;
import java.util.NoSuchElementException;

public class SimpleLinkedList<T> {

    private Node head;
    private int size;

    public SimpleLinkedList() {
        this.size = 0;
    }

    public SimpleLinkedList(final T[] values) {
        for (int i = values.length - 1; i >= 0; i--) {
            push(values[i]);
        }
    }

    public int size() {
        return size;
    }

    public T pop() {
        if (head == null) {
            throw new NoSuchElementException();
        }
        T value = head.value;
        head = head.next;
        size--;
        return value;
    }

    public void push(final T value) {
        head = new Node(value, head);
        size++;
    }

    public void reverse() {
        if (head == null || head.next == null) {
            return;
        }

        Node prev = null;
        Node current = head;
        Node next = null;

        while (current != null) {
            next = current.next;
            current.next = prev;
            prev = current;
            current = next;
        }
        head = prev;
    }

    public T[] asArray(final Class<T> clazz) {
        T[] array = (T[]) Array.newInstance(clazz, size);
        Node node = head;
        for (int i = 0; i < size; i++) {
            array[i] = node.value;
            node = node.next;
        }
        return array;
    }

    private class Node {

        private final T value;
        private Node next;

        private Node(final T value, final Node next) {
            this.value = value;
            this.next = next;
        }
    }
}